lib/class-wp-rest-template-parts-controller.php

```diff
@@ -0,0 +1,130 @@
+<?php
+/**
+ * REST API: WP_REST_Template_Parts_Controller class
+ *
+ * @subpackage REST_API
+ * @package    WordPress
+ */
+
+/**
+ * Creates a template part auto-draft if it doesn't exist yet.
+ *
+ * @access private
+ *
+ * @param string $slug    Template part slug.
+ * @param string $theme   Template part theme.
+ * @param string $content Template part content.
+ */
+function create_auto_draft_for_template_part_block( $slug, $theme, $content ) {
+	// A published post might already exist if this template part
+	// was customized elsewhere or if it's part of a customized
+	// template. We also check if an auto-draft was already created
+	// because preloading can make this run twice, so, different code
+	// paths can end up with different posts for the same template part.
+	// E.g. The server could send back post ID 1 to the client, preload,
+	// and create another auto-draft. So, if the client tries to resolve the
+	// post ID from the slug and theme, it won't match with what the server sent.
+	$template_part_query = new WP_Query(
+		array(
+			'post_type'      => 'wp_template_part',
+			'post_status'    => array( 'publish', 'auto-draft' ),
```

Comment:
I wonder if we should even check `publish` here.  The published version would be customized by the user, and creating an auto-draft anyways in this case could be a way to go back to the theme's default.